### PR TITLE
sql: add VIEWACTIVITY and CANCELQUERY role options

### LINKED ISSUE #52578: sql: add CANCELQUERY role option
This is a new role option which allows a non-admin user to cancel queries of other users. Without this privilege, users can only cancel their own queries. Even with this privilege, non-admins cannot cancel admin queries.

### LINKED ISSUE #52579: sql: add CANCELSESSION role option
This is a new role option which allows users to cancel sessions of other users. Without this privilege, users can only cancel their own session. Even with this privilege, non-admins cannot cancel admin sessions.

### LINKED ISSUE #53277: sql: non-root admins cannot cancel queries or sessions
Our docs claim that any admin user can run CANCEL QUERY and CANCEL SESSION against other users' queries and sessions. This is not the case. We currently only permit the `root` user to do this.

### PR DESCRIPTION
**Pull request by solongordon**
Release note (sql change): Introduced a new VIEWACTIVITY role option.
This grants non-admin roles the ability to see other users' sessions and
queries through the following means:
- SHOW SESSIONS
- SHOW QUERIES
- the WebUI Statements page

Release note (sql change): Introduced a new CANCELQUERY role option.
This grants non-admin roles the ability to cancel other users' queries
and sessions. Note that non-admins are not allowed to cancel admins'
queries or sessions.

Release note (bug fix): Admin users are now permitted to cancel other
users' queries and sessions. Previously only the root user was allowed
to do so.

Fixes #52578
Fixes #52579
Fixes #53277

### CODE REVIEW

File: pkg/server/admin.go [hunk 1]
```diff
@@ -2441,4 +2463,33 @@ func (s *adminServer) hasAdminRole(ctx context.Context, sessionUser string) (boo
...
+		return true, nil
+	}
+	rows, cols, err := s.server.sqlServer.internalExecutor.QueryWithCols(
+		ctx, "check-role-option", nil, /* txn */
+		sessiondata.InternalExecutorOverride{User: user},
+		"SELECT crdb_internal.has_role_option($1)", roleOption.String())
+	if err != nil {
+		return false, err
+	}
+	if len(rows) != 1 || len(cols) != 1 {
```
Comment (rohany) [inline pkg/server/admin.go:2480]: why do you need the columns here?

File: pkg/server/admin.go [hunk 2]
```diff
@@ -2441,4 +2463,33 @@ func (s *adminServer) hasAdminRole(ctx context.Context, sessionUser string) (boo
...
 
-var errInsufficientPrivilege = status.Error(codes.PermissionDenied, "this operation requires admin privilege")
+func (s *adminServer) hasRoleOption(
+	ctx context.Context, user string, roleOption roleoption.Option,
+) (bool, error) {
+	if user == security.RootUser {
+		// Shortcut.
+		return true, nil
+	}
+	rows, cols, err := s.server.sqlServer.internalExecutor.QueryWithCols(
```
Comment (rohany) [inline pkg/server/admin.go:2473]: Is it possible to share these queries with the planner, maybe helper methods that take in an ie interface?

File: pkg/server/status.go [hunk 3]
```diff
@@ -1866,8 +1910,49 @@ func (s *statusServer) CancelQuery(
...
 	}
 
+	reqUserHasAdmin, err := s.admin.hasAdminRole(ctx, req.Username)
+	if err != nil {
+		return nil, err
+	}
+
+	if !reqUserHasAdmin {
+		// Check if the user has permission to see the query's session.
+		var session serverpb.Session
```
Comment (rohany) [inline pkg/server/status.go:1920]: this code seems duplicated with the above

**Review by rohany — CHANGES_REQUESTED** [review]
Will take a closer look again, preliminary comments

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/server/admin.go, line 2473 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFRgtnKMD9YrGiAZEQa-r2-2473:-MFSXE0v1MLY2v5iBA0T:b-qeqkf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7281832908dca586ce26dc9e855819a7538b05af/pkg/server/admin.go#L2473)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Is it possible to share these queries with the planner, maybe helper methods that take in an ie interface?
</blockquote></details>

Can you say more? The planner already has a `HasRoleOption` method, which is how the `crdb_internal.has_role_option` built-in is implemented. I used the built-in here since there's no planner available.
___
*[pkg/server/admin.go, line 2480 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFRgtnKMD9YrGiAZEQa-r2-2480:-MFSWoNA7mgYwbU8AK0x:b-soty14) ([raw file](https://github.com/cockroachdb/cockroach/blob/7281832908dca586ce26dc9e855819a7538b05af/pkg/server/admin.go#L2480)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

why do you need the columns here?
</blockquote></details>

This was cargo-culted from `hasAdminRole` but it does seem like a good idea to assert that the result shape is as expected. I'm going to switch this to ignore `cols` and check `len(rows[0]) == 1` instead.
___
*[pkg/server/status.go, line 1920 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFRgtnKMD9YrGiAZEQd-r2-1920:-MFSW-fB0uszpI4j6RmL:b-43pp6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7281832908dca586ce26dc9e855819a7538b05af/pkg/server/status.go#L1920)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

this code seems duplicated with the above
</blockquote></details>

Might be able to factor something out but this code is matching on query ID rather than session ID above.

**Review by rohany — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/server/admin.go, line 2473 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFRgtnKMD9YrGiAZEQa-r2-2473:-MFT2LDK--yNwiX7QwhM:b-wow048) ([raw file](https://github.com/cockroachdb/cockroach/blob/7281832908dca586ce26dc9e855819a7538b05af/pkg/server/admin.go#L2473)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Can you say more? The planner already has a `HasRoleOption` method, which is how the `crdb_internal.has_role_option` built-in is implemented. I used the built-in here since there's no planner available.
</blockquote></details>

Never mind. I thought the planner's implementation also used an IE to make a query like this, but thats not true
___
*[pkg/server/status.go, line 1920 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFRgtnKMD9YrGiAZEQd-r2-1920:-MFT2efG8SCG3OtQ8qMV:b-chstyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7281832908dca586ce26dc9e855819a7538b05af/pkg/server/status.go#L1920)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Might be able to factor something out but this code is matching on query ID rather than session ID above.
</blockquote></details>

Yeah. Perhaps the non-admin check logic could be pulled out, but thats it.
___
*[pkg/sql/internal_test.go, line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFT37Qn0j0ClAVP7ZTi:-MFT37Qn0j0ClAVP7ZTj:b-lnc8qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcdaaf7fbc602f1451b4a11488227bf7578b4/pkg/sql/internal_test.go#L200)):*
> ```Go
> }
> 
> func TestQueryHasRoleOptionWithNoTxn(t *testing.T) {
> ```

What's the goal of this test?

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @solongordon)

___
*[pkg/sql/internal_test.go, line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MFT37Qn0j0ClAVP7ZTi:-MFTKmQ65cc31MnZUNb4:b44gsmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2fdcdaaf7fbc602f1451b4a11488227bf7578b4/pkg/sql/internal_test.go#L200)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

What's the goal of this test?
</blockquote></details>

Just testing the new `crdb_internal.has_role_option` built-in. Now that I think about it this might be nicer as a logic test, but I was mimicking the existing `is_admin` unit test.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @solongordon)

___
*[pkg/server/status.go, line 1894 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MG3chDeEXp8b2dNDiM3:-MG3chDeEXp8b2dNDiM4:bv704xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fcf5e4d78afcf16c2677ae65fe2d48bb52e0ac9/pkg/server/status.go#L1894)):*
> ```Go
> 	if !isAdmin && sessionUser != req.Username {
> 		// A user can only cancel their own queries.
> 		return nil, errRequiresAdmin
> ```

Drive-by comment since I'm fixing some merge conflicts with this PR in another PR. Is it correct for this check to stay in? I assume that this will ignore the `CANCELQUERY` privilege if a non-admin user with this privilege tries to cancel another user's query.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @solongordon)

___
*[pkg/server/status.go, line 1894 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53291#-MG3chDeEXp8b2dNDiM3:-MG3ta0A6HdZaANnXUOz:b-kxwcxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fcf5e4d78afcf16c2677ae65fe2d48bb52e0ac9/pkg/server/status.go#L1894)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Drive-by comment since I'm fixing some merge conflicts with this PR in another PR. Is it correct for this check to stay in? I assume that this will ignore the `CANCELQUERY` privilege if a non-admin user with this privilege tries to cancel another user's query.
</blockquote></details>

I think the check is still correct but the comment is misleading. When a CANCEL QUERY statement is run, this request is invoked with `root` and the user who issued the query is specified in `req.Username`. However if this API is invoked from the Admin UI, `sessionUser` and `req.Username` should match. I'll have to check if this is actually the case in @jordanlewis's outstanding PR for the Sessions page.

**Pull request merged**